File: ts/src/bind-handler.ts
import { Buffer } from "node:buffer";

import type { Context, Handler } from "./context.js";
import { AppError, AppTheoryError } from "./errors.js";
import { json } from "./response.js";
import type { ValidationRuleSpec, ValidationSchema } from "./validate.js";
import { validateOrThrow } from "./validate.js";

export type BindSource = "body" | "query" | "path" | "header";
export type BindFieldType = "string" | "int" | "bool" | "float" | "duration";

export interface BindFieldSpec {
  source: BindSource;
  name?: string;
  type?: BindFieldType;
  array?: boolean;
  field?: string;
  validate?: readonly ValidationRuleSpec[];
}

export type BindFieldSpecs<Req> = Partial<
  Record<Extract<keyof Req, string>, BindFieldSpec>
>;

export interface BindConfig<Req> {
  body?: boolean;
  query?: boolean;
  path?: boolean;
  headers?: boolean;
  strictJson?: boolean;
  successStatus?: number;
  fields?: BindFieldSpecs<Req>;
  validation?: ValidationSchema<Req>;
  validate?: (ctx: Context, req: Req) => void | Promise<void>;
}

interface SourceValueResult {
  present: boolean;
  values: unknown[];
}

const MAX_PORTABLE_INTEGER = Number.MAX_SAFE_INTEGER;

export type TypedHandler<Req, Resp> = (
  ctx: Context,
  req: Req,
) => Resp | Promise<Resp>;

export function bindHandler<Req, Resp>(
  config: BindConfig<Req>,
  handler: TypedHandler<Req, Resp>,
): Handler {
  return async (ctx: Context) => {
    const req = await bindRequest(ctx, config);
    const resp = await handler(ctx, req);
    return json(config.successStatus ?? 200, resp);
  };
}

export async function bindRequest<Req>(
  ctx: Context,
  config: BindConfig<Req>,
): Promise<Req> {
  const fields = (config.fields ?? {}) as BindFieldSpecs<Req>;
  const fieldKeys = Object.keys(fields) as Extract<keyof Req, string>[];
  let req: Record<string, unknown> = {};
  let bodyValue: Record<string, unknown> | null = null;

  if (config.body) {
    bodyValue = parseBody(ctx);
    const bodyFieldNames = new Set(
      fieldKeys
        .filter((key) => fields[key]?.source === "body")
        .map((key) => fields[key]?.name ?? key),
    );
    if (config.strictJson) {
      for (const key of Object.keys(bodyValue)) {
        if (!bodyFieldNames.has(key)) {
          throw bindingError("body", key, "", undefined);
        }
      }
    }
    if (fieldKeys.length === 0) {
      req = { ...bodyValue };
    }
  }

  for (const key of fieldKeys) {
    const spec = fields[key];
    if (!spec) continue;
    const name = spec.name ?? key;
    const source = sourceValues(ctx, bodyValue, spec.source, name);
    if (!source.present) continue;
    try {
      req[key] = spec.array
        ? source.values.map((value) => parseValue(value, spec.type ?? "string"))
        : parseValue(
            source.values.length > 0 ? source.values[0] : "",
            spec.type ?? "string",
          );
    } catch (err) {
      throw bindingError(spec.source, name, spec.field ?? key, err);
    }
  }

  const validation = mergeValidation(config.validation, fields);
  validateOrThrow(req as Req, validation);

  if (typeof config.validate === "function") {
    try {
      await config.validate(ctx, req as Req);
    } catch (err) {
      throw normalizeValidationError(err);
    }
  }

  return req as Req;
}

function parseBody(ctx: Context): Record<string, unknown> {
  const body = Buffer.from(ctx.request.body ?? []);
  if (body.length === 0) {
    throw new AppTheoryError("app.bad_request", "request body is empty", {
      statusCode: 400,
    });
  }
  try {
    const parsed = JSON.parse(body.toString("utf8")) as unknown;
    if (!isRecord(parsed)) {
      throw new TypeError("request body must be a JSON object");
    }
    return parsed;
  } catch (err) {
    if (err instanceof AppTheoryError) throw err;
    throw new AppTheoryError("app.bad_request", "invalid json", {
      statusCode: 400,
      cause: err,
    });
  }
}

function sourceValues(
  ctx: Context,
  bodyValue: Record<string, unknown> | null,
  source: BindSource,
  name: string,
): SourceValueResult {
  switch (source) {
    case "body": {
      if (
        !bodyValue ||
        !Object.prototype.hasOwnProperty.call(bodyValue, name)
      ) {
        return { present: false, values: [] };
      }
      const value = bodyValue[name];
      if (Array.isArray(value)) return { present: true, values: value };
      return { present: true, values: [value] };
    }
    case "query": {
      const values = ctx.request.query?.[name];
      return values && values.length > 0
        ? { present: true, values: values.map((value) => String(value)) }
        : { present: false, values: [] };
    }
    case "path": {
      const value = ctx.params?.[name];
      return value === undefined
        ? { present: false, values: [] }
        : { present: true, values: [String(value)] };
    }
    case "header": {
      const values = ctx.request.headers?.[name.toLowerCase()];
      return values && values.length > 0
        ? { present: true, values: values.map((value) => String(value)) }
        : { present: false, values: [] };
    }
  }
}

function parseValue(raw: unknown, type: BindFieldType): unknown {
  switch (type) {
    case "string":
      return raw === null ? null : String(raw);
    case "int": {
      const rawText = String(raw ?? "");
      if (!/^[+-]?\d+$/.test(rawText)) throw new Error("invalid integer");
      const value = Number.parseInt(rawText, 10);
      if (
        !Number.isSafeInteger(value) ||
        Math.abs(value) > MAX_PORTABLE_INTEGER
      ) {
        throw new Error("integer outside portable safe range");
      }
      return value;
    }
    case "bool": {
      const normalized = String(raw ?? "")
        .trim()
        .toLowerCase();
      if (["1", "t", "true"].includes(normalized)) return true;
      if (["0", "f", "false"].includes(normalized)) return false;
      throw new Error("invalid boolean");
    }
    case "float": {
      const rawText = String(raw ?? "");
      if (!/^[+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?$/.test(rawText)) {
        throw new Error("invalid float");
      }
      const value = Number(rawText);
      if (!Number.isFinite(value)) throw new Error("invalid float");
      return value;
    }
    case "duration":
      return parseDuration(String(raw ?? ""));
  }
}

function parseDuration(raw: string): string {
  const trimmed = raw.trim();
  if (!trimmed) throw new Error("invalid duration");
  let sign = 1n;
  let body = trimmed;
  if (body[0] === "-" || body[0] === "+") {
    sign = body[0] === "-" ? -1n : 1n;
    body = body.slice(1);
  }
  if (!body) throw new Error("invalid duration");

  const segments = parseDurationSegments(body);
  let totalNs = 0n;
  for (const segment of segments) {
    totalNs += decimalDurationToNs(segment.amount, segment.unit);
  }
  totalNs *= sign;
  if (totalNs % 1000n !== 0n) {
    throw new Error("duration precision below one microsecond is not portable");
  }
  return formatDurationNs(totalNs);
}

function durationUnitNs(unit: string): bigint {
  switch (unit) {
    case "h":
      return 3_600_000_000_000n;
    case "m":
      return 60_000_000_000n;
    case "s":
      return 1_000_000_000n;
    case "ms":
      return 1_000_000n;
    case "us":
    case "µs":
      return 1_000n;
    case "ns":
      return 1n;
    default:
      throw new Error("invalid duration");
  }
}

interface DurationSegment {
  readonly amount: string;
  readonly unit: string;
}

function parseDurationSegments(body: string): DurationSegment[] {
  const segments: DurationSegment[] = [];
  let index = 0;
  while (index < body.length) {
    const amountStart = index;
    let sawWholeDigit = false;
    while (index < body.length && isAsciiDigit(body.charCodeAt(index))) {
      sawWholeDigit = true;
      index += 1;
    }

    if (index < body.length && body[index] === ".") {
      index += 1;
      const fractionStart = index;
      while (index < body.length && isAsciiDigit(body.charCodeAt(index))) {
        index += 1;
      }
      if (!sawWholeDigit && index === fractionStart) {
        throw new Error("invalid duration");
      }
    } else if (!sawWholeDigit) {
      throw new Error("invalid duration");
    }

    const amount = body.slice(amountStart, index);
    const unit = readDurationUnit(body, index);
    if (!unit) {
      throw new Error("invalid duration");
    }
    index += unit.length;
    segments.push({ amount, unit });
  }
  if (segments.length === 0) {
    throw new Error("invalid duration");
  }
  return segments;
}

function readDurationUnit(body: string, index: number): string {
  for (const unit of ["ns", "us", "µs", "ms", "s", "m", "h"]) {
    if (body.startsWith(unit, index)) {
      return unit;
    }
  }
  return "";
}

function decimalDurationToNs(amount: string, unit: string): bigint {
  const unitNs = durationUnitNs(unit);
  const pieces = amount.split(".");
  const wholeRaw = pieces[0] ?? "";
  const fractionRaw = pieces[1] ?? "";
  const whole = wholeRaw === "" ? 0n : BigInt(wholeRaw);
  let out = whole * unitNs;
  if (fractionRaw !== "") {
    const fraction = BigInt(fractionRaw);
    const scale = 10n ** BigInt(fractionRaw.length);
    out += (fraction * unitNs) / scale;
  }
  return out;
}

function formatDurationNs(totalNs: bigint): string {
  const sign = totalNs < 0n ? "-" : "";
  let remaining = totalNs < 0n ? -totalNs : totalNs;
  const hour = 3_600_000_000_000n;
  const minute = 60_000_000_000n;
  const second = 1_000_000_000n;
  const millisecond = 1_000_000n;
  const microsecond = 1_000n;
  if (remaining === 0n) return "0s";

  const parts: string[] = [];
  if (remaining >= second) {
    const hours = remaining / hour;
    if (hours > 0n) {
      parts.push(`${hours}h`);
      remaining %= hour;
      const minutes = remaining / minute;
      parts.push(`${minutes}m`);
      remaining %= minute;
    } else {
      const minutes = remaining / minute;
      if (minutes > 0n) {
        parts.push(`${minutes}m`);
        remaining %= minute;
      }
    }
    const seconds = remaining / second;
    const fraction = remaining % second;
    parts.push(formatDecimalUnit(seconds, fraction, 9, "s"));
    return `${sign}${parts.join("")}`;
  }

  if (remaining >= millisecond) {
    return `${sign}${formatDecimalUnit(
      remaining / millisecond,
      remaining % millisecond,
      6,
      "ms",
    )}`;
  }
  if (remaining >= microsecond) {
    return `${sign}${formatDecimalUnit(
      remaining / microsecond,
      remaining % microsecond,
      3,
      "µs",
    )}`;
  }
  return `${sign}${remaining}ns`;
}

function formatDecimalUnit(
  whole: bigint,
  fraction: bigint,
  fractionWidth: number,
  unit: string,
): string {
  if (fraction === 0n) return `${whole}${unit}`;
  let text = fraction.toString().padStart(fractionWidth, "0");
  text = trimTrailingZeros(text);
  return `${whole}.${text}${unit}`;
}

function trimTrailingZeros(value: string): string {
  let end = value.length;
  while (end > 0 && value.charCodeAt(end - 1) === 48) {
    end -= 1;
  }
  return value.slice(0, end);
}

function isAsciiDigit(code: number): boolean {
  return code >= 48 && code <= 57;
}

function bindingError(
  source: BindSource,
  name: string,
  field: string,
  cause: unknown,
): AppTheoryError {
  const message = field
    ? `invalid ${source} binding for ${field}`
    : `invalid ${source} binding: ${name}`;
  const details: Record<string, unknown> = { source, name };
  if (field) details["field"] = field;
  return new AppTheoryError("app.bad_request", message, {
    statusCode: 400,
    details,
    cause,
  });
}

function normalizeValidationError(err: unknown): AppTheoryError {
  if (err instanceof AppTheoryError) {
    if (err.code === "app.validation_failed" && !err.statusCode) {
      err.statusCode = 422;
    }
    return err;
  }
  if (err instanceof AppError) {
    const options =
      err.code === "app.validation_failed"
        ? { statusCode: 422, cause: err }
        : { cause: err };
    return new AppTheoryError(err.code, err.message, options);
  }
  return new AppTheoryError("app.validation_failed", "validation failed", {
    statusCode: 422,
    cause: err,
  });
}

function mergeValidation<Req>(
  validation: ValidationSchema<Req> | undefined,
  fields: BindFieldSpecs<Req>,
): ValidationSchema<Req> | undefined {
  const merged: Record<string, ValidationRuleSpec[]> = {};
  for (const [key, rules] of Object.entries(
    (validation as Record<string, readonly ValidationRuleSpec[]>) ?? {},
  )) {
    const spec = (fields as Record<string, BindFieldSpec>)[key];
    const wireName = spec?.name ?? key;
    merged[key] = [...(rules ?? [])].map((rule) =>
      withCanonicalValidationField(rule, wireName),
    );
  }
  for (const [key, spec] of Object.entries(
    fields as Record<string, BindFieldSpec>,
  )) {
    if (!spec.validate) continue;
    const wireName = spec.name ?? key;
    merged[key] = [
      ...(merged[key] ?? []),
      ...spec.validate.map((rule) =>
        withCanonicalValidationField(rule, wireName),
      ),
    ];
  }
  return Object.keys(merged).length > 0
    ? (merged as ValidationSchema<Req>)
    : undefined;
}

function withCanonicalValidationField(
  rule: ValidationRuleSpec,
  field: string,
): ValidationRuleSpec {
  return rule.field === undefined ? { ...rule, field } : { ...rule };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

File: py/src/apptheory/bind_handler.py
from __future__ import annotations

import asyncio
import datetime as dt
import inspect
import json as jsonlib
import re
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field, fields, is_dataclass
from types import NoneType, UnionType
from typing import Any, get_args, get_origin

from apptheory.context import Context
from apptheory.errors import AppError, AppTheoryError, app_theory_error_from_app_error, status_for_error_code
from apptheory.response import Response, json
from apptheory.validate import ValidationRule, ValidationSchema, validate_or_raise

_UNSET = object()


@dataclass(frozen=True, slots=True)
class BindField:
    source: str
    name: str = ""
    value_type: str = ""
    field_name: str = ""
    array: bool = False
    validate: list[ValidationRule] | None = None


@dataclass(frozen=True, slots=True)
class _SourceValues:
    present: bool
    values: list[Any]


@dataclass(slots=True)
class BindConfig[ReqT]:
    model: type[ReqT]
    body: bool = False
    query: bool = False
    path: bool = False
    headers: bool = False
    strict_json: bool = False
    success_status: int = 200
    validation: ValidationSchema | None = None
    validate: Callable[[Context, ReqT], None | Awaitable[None]] | None = None


type TypedHandler[ReqT, RespT] = Callable[[Context, ReqT], RespT | Awaitable[RespT]]


def body(
    name: str = "",
    *,
    default: Any = _UNSET,
    value_type: str = "",
    validate: list[ValidationRule] | None = None,
    array: bool = False,
):
    return _bind_field("body", name, default=default, value_type=value_type, validate=validate, array=array)


def query(
    name: str = "",
    *,
    default: Any = _UNSET,
    value_type: str = "",
    validate: list[ValidationRule] | None = None,
    array: bool = False,
):
    return _bind_field("query", name, default=default, value_type=value_type, validate=validate, array=array)


def path(
    name: str = "",
    *,
    default: Any = _UNSET,
    value_type: str = "",
    validate: list[ValidationRule] | None = None,
    array: bool = False,
):
    return _bind_field("path", name, default=default, value_type=value_type, validate=validate, array=array)


def header(
    name: str = "",
    *,
    default: Any = _UNSET,
    value_type: str = "",
    validate: list[ValidationRule] | None = None,
    array: bool = False,
):
    return _bind_field("header", name, default=default, value_type=value_type, validate=validate, array=array)


def bind_handler[ReqT, RespT](config: BindConfig[ReqT], handler: TypedHandler[ReqT, RespT]):
    def _handler(ctx: Context) -> Response:
        req = bind_request(ctx, config)
        resp = _resolve(handler(ctx, req))
        return json(config.success_status or 200, resp)

    return _handler


def bind_request[ReqT](ctx: Context, config: BindConfig[ReqT]) -> ReqT:
    model = config.model
    body_value: dict[str, Any] | None = None
    if config.body:
        body_value = _parse_body(ctx)
    bind_fields = _model_fields(model, body_enabled=config.body)
    if config.strict_json and body_value is not None:
        expected_body_names = {bind.name for bind, _annotation in bind_fields.values() if bind.source == "body"}
        for key in body_value:
            if key not in expected_body_names:
                raise _binding_error("body", key, "", None)

    values: dict[str, Any] = {}
    for attr, (bind, annotation) in bind_fields.items():
        source = _source_values(ctx, body_value, bind.source, bind.name)
        if not source.present:
            continue
        try:
            values[attr] = _convert_values(source.values, annotation, bind)
        except Exception as exc:
            raise _binding_error(bind.source, bind.name, bind.field_name or attr, exc) from exc

    req = _construct_model(model, values)
    validation = _merge_validation(config.validation, bind_fields)
    validate_or_raise(req, validation)

    if config.validate is not None:
        try:
            _resolve(config.validate(ctx, req))
        except Exception as exc:
            raise _normalize_validation_error(exc) from exc

    return req


def _bind_field(
    source: str,
    name: str,
    *,
    default: Any,
    value_type: str,
    validate: list[ValidationRule] | None,
    array: bool = False,
):
    metadata = {"apptheory_bind": BindField(source, name, value_type=value_type, validate=validate, array=array)}
    if default is _UNSET:
        return field(default=None, metadata=metadata)
    return field(default=default, metadata=metadata)


def _resolve[RespT](value: RespT | Awaitable[RespT]) -> RespT:
    if not inspect.isawaitable(value):
        return value

    async def _await_any(awaitable: Awaitable[RespT]) -> RespT:
        return await awaitable

    try:
        return asyncio.run(_await_any(value))
    except RuntimeError as exc:  # pragma: no cover
        raise RuntimeError("apptheory: cannot resolve awaitable from sync bind handler") from exc


def _parse_body(ctx: Context) -> dict[str, Any]:
    if not ctx.request.body:
        raise AppTheoryError("app.bad_request", "request body is empty", status_code=400)
    try:
        parsed = jsonlib.loads(ctx.request.body.decode("utf-8"))
    except Exception as exc:
        raise AppTheoryError("app.bad_request", "invalid json", status_code=400, cause=exc) from exc
    if not isinstance(parsed, dict):
        raise AppTheoryError("app.bad_request", "invalid json", status_code=400)
    return parsed


def _model_fields(model: type[Any], *, body_enabled: bool) -> dict[str, tuple[BindField, Any]]:
    out: dict[str, tuple[BindField, Any]] = {}
    if is_dataclass(model):
        for dataclass_field in fields(model):
            bind = dataclass_field.metadata.get("apptheory_bind")
            if bind is None:
                if not body_enabled:
                    continue
                bind = BindField("body", dataclass_field.name)
            if not bind.name:
                bind = BindField(
                    bind.source,
                    dataclass_field.name,
                    value_type=bind.value_type,
                    field_name=bind.field_name or dataclass_field.name,
                    array=bind.array,
                    validate=bind.validate,
                )
            elif not bind.field_name:
                bind = BindField(
                    bind.source,
                    bind.name,
                    value_type=bind.value_type,
                    field_name=dataclass_field.name,
                    array=bind.array,
                    validate=bind.validate,
                )
            out[dataclass_field.name] = (bind, dataclass_field.type)
        return out

    annotations = getattr(model, "__annotations__", {}) or {}
    for attr, annotation in annotations.items():
        if body_enabled:
            out[attr] = (BindField("body", str(attr), field_name=str(attr)), annotation)
    return out


def _source_values(ctx: Context, body_value: dict[str, Any] | None, source: str, name: str) -> _SourceValues:
    if source == "body":
        if body_value is None or name not in body_value:
            return _SourceValues(False, [])
        value = body_value.get(name)
        return _SourceValues(True, list(value) if isinstance(value, list) else [value])
    if source == "query":
        values = list((ctx.request.query or {}).get(name) or [])
        return _SourceValues(bool(values), values)
    if source == "path":
        value = (ctx.params or {}).get(name)
        return _SourceValues(False, []) if value is None else _SourceValues(True, [value])
    if source == "header":
        values = list((ctx.request.headers or {}).get(name.lower()) or [])
        return _SourceValues(bool(values), values)
    return _SourceValues(False, [])


def _construct_model[ReqT](model: type[ReqT], values: dict[str, Any]) -> ReqT:
    return model(**values)


def _convert_values(values: list[Any], annotation: Any, bind: BindField) -> Any:
    inner, is_list = _inner_type(annotation)
    if bind.array or is_list:
        return [_convert_one(value, inner, bind.value_type) for value in values]
    return _convert_one(values[0], inner, bind.value_type)


def _inner_type(annotation: Any) -> tuple[Any, bool]:
    origin = get_origin(annotation)
    args = [arg for arg in get_args(annotation) if arg is not NoneType]
    if origin is UnionType and args:
        return _inner_type(args[0])
    if origin in {list, tuple, set} and args:
        return args[0], True
    return annotation, False


def _convert_one(value: Any, annotation: Any, value_type: str) -> Any:
    target = value_type or _type_name(annotation)
    if target in {"str", "string"}:
        if value is None:
            return None
        return str(value)
    if target in {"int", "integer"}:
        raw = str(value)
        if re.fullmatch(r"[+-]?\d+", raw) is None:
            raise ValueError("invalid integer")
        return int(raw)
    if target in {"bool", "boolean"}:
        raw = str(value).strip().lower()
        if raw in {"1", "t", "true"}:
            return True
        if raw in {"0", "f", "false"}:
            return False
        raise ValueError("invalid boolean")
    if target in {"float", "number"}:
        return float(str(value))
    if target in {"duration", "timedelta"} or annotation is dt.timedelta:
        return _parse_duration(str(value))
    return value


def _type_name(annotation: Any) -> str:
    if annotation is str:
        return "str"
    if annotation is int:
        return "int"
    if annotation is bool:
        return "bool"
    if annotation is float:
        return "float"
    if annotation is dt.timedelta:
        return "timedelta"
    return str(annotation)


def _parse_duration(raw: str) -> dt.timedelta:
    units = {"h": 3600, "m": 60, "s": 1, "ms": 0.001, "us": 0.000001, "µs": 0.000001, "ns": 0.000000001}
    pattern = re.compile(r"([+-]?\d+(?:\.\d+)?)(ns|us|µs|ms|s|m|h)")
    consumed = ""
    seconds = 0.0
    for match in pattern.finditer(raw.strip()):
        consumed += match.group(0)
        seconds += float(match.group(1)) * units[match.group(2)]
    if not raw.strip() or consumed != raw.strip():
        raise ValueError("invalid duration")
    return dt.timedelta(seconds=seconds)


def format_duration(value: dt.timedelta) -> str:
    total_ns = int(value.total_seconds() * 1_000_000_000)
    sign = "-" if total_ns < 0 else ""
    remaining = abs(total_ns)
    hour = 3_600_000_000_000
    minute = 60_000_000_000
    second = 1_000_000_000
    parts: list[str] = []
    hours, remaining = divmod(remaining, hour)
    if hours:
        parts.append(f"{hours}h")
    minutes, remaining = divmod(remaining, minute)
    if minutes:
        parts.append(f"{minutes}m")
    seconds, remaining = divmod(remaining, second)
    if seconds or parts:
        parts.append(f"{seconds}s")
    elif remaining == 0:
        parts.append("0s")
    if remaining:
        parts.append(f"{remaining}ns")
    return sign + "".join(parts)


def _binding_error(source: str, name: str, field_name: str, cause: Exception | None) -> AppTheoryError:
    message = f"invalid {source} binding for {field_name}" if field_name else f"invalid {source} binding: {name}"
    details: dict[str, Any] = {"source": source, "name": name}
    if field_name:
        details["field"] = field_name
    return AppTheoryError("app.bad_request", message, status_code=400, details=details, cause=cause)


def _merge_validation(
    config_validation: ValidationSchema | None, bind_fields: dict[str, tuple[BindField, Any]]
) -> ValidationSchema | None:
    merged: ValidationSchema = {key: list(value) for key, value in (config_validation or {}).items()}
    for attr, (bind, _annotation) in bind_fields.items():
        if not bind.validate:
            continue
        rules = [ValidationRule(rule.rule, rule.value, field=bind.name, message=rule.message) for rule in bind.validate]
        merged.setdefault(attr, []).extend(rules)
    return merged or None


def _normalize_validation_error(exc: Exception) -> AppTheoryError:
    if isinstance(exc, AppTheoryError):
        if exc.code == "app.validation_failed" and not exc.status_code:
            exc.status_code = 422
        return exc
    if isinstance(exc, AppError):
        return app_theory_error_from_app_error(exc).with_status_code(status_for_error_code(exc.code)).with_cause(exc)
    return AppTheoryError("app.validation_failed", "validation failed", status_code=422, cause=exc)
